### DISCUSSION TITLE
FreeBSD: kernel-pfroute plugin not making use of the RTF_GATEWAY flag in case of a routing entry via gateway

### DISCUSSION THREAD

**Discussion by pa5h1nh0** [strongswan/strongswan#2517] [Q&A]
I'm just curious for what might be the reason, even if it's not actually a problem cuz it still just works. I found [this same question](https://users.strongswan.narkive.com/0qJvttEE/strongswan-strongswan-routing) on a mailing list from 10 years ago, but without an answer and no follow up. If I get this log from strongswan `installing route: 192.168.200.0/24 via 192.168.68.40 src 192.168.201.1 dev eth0`, as a user I'd expect to see the following entry in the routing table:
```
# nestat -rn4
Routing tables

Internet:
Destination        Gateway            Flags     Netif Expire
...
192.168.200.0/24   192.168.68.40      UGS       eth0
```

instead what I'm actually seeing is this:
```
# nestat -rn4
Routing tables

Internet:
Destination        Gateway            Flags     Netif Expire
...
192.168.200.0/24   link#3             US        eth0
```
i.e. the gateway specified in the log is missing, the same is for the missing flag `G`. Looking at the code in `kernel_pfrout_net.c` file, indeed I notice that even though a gateway is specified, the `RTF_GATEWAY` is still not set in `rt_msghdr::rtm_flags`. Is there a reason for not setting the `RTF_GATEWAY` flag even though it's explicitly asked to add a gateway? Why adding the gateway IP in the first place then and send the request with the missing flag over the `PF_ROUTE` socket if it gets totally ignored by the kernel anyway? Wouldn't all this mean that it is _technically_ a bug, but apparently not _functionally_ cuz no one is complaining as it still _just works TM?_

**Comment by tobiasbrunner** [comment]
Honestly, no idea. It's been like this since route installation was added to that plugin over 11 year ago (I think that was for macOS, not FreeBSD originally). Although interface names were initially preferred over gateways (both set via `RTA_GATEWAY`), which was changed shortly afterwards (to pass the interface separately with `RTA_IFP`). And I recently added a change that configures the interface name again via `RTA_GATEWAY` if no gateway is known or it's for a different address family, which seems to have an effect even without `RTF_GATEWAY` (see #2465). Since it works, I'd be reluctant to change anything due to possible negative side-effects.
